cogs/scheduledmessage.py

```diff
@@ -0,0 +1,22 @@
+from discord.ext import tasks, commands
+
+class ScheduledMessage(commands.Cog):
+    def __init__(self, bot: commands.Bot, message: str, channel_id: int, interval_seconds: int):
+        self.bot = bot
+        self.message = message
+        self.channel_id = channel_id
+        self.interval_seconds = interval_seconds
+        # decorate printer and before_printer to allow dynamic interval be passed to tasks.loop
+        self.printer = tasks.loop(seconds=self.interval_seconds)(self.printer)
```

Comment:
Does the decorator not work if we do it this way?

```python
class ScheduledMessage(commands.Cog):
   def __init__(self, ..):
    ..
    self.printer.start()

  @tasks.loop(seconds=self.interval_seconds)
  async def printer(self):
        channel = self.bot.get_channel(self.channel_id)
        await channel.send(self.message)
```

Reply:
Not entirely sure being newer to decorators in python but I had it that way initially and it failed because the class wasn't initialized yet. From googling it seems decorator binds to printer() before __init__ is ran so self( and self.interval_seconds) isn't defined yet. I used the 4th answer from https://stackoverflow.com/questions/11731136/class-method-decorator-with-self-arguments to do it this way as other answers required being able to modify the decorator. This is the extent of what I know about decorators so I don't know any other way to get it to work.

Reply:
Yeah you're totally right! To pass the `interval_seconds` we may need to make some modifications. I have a gist with some example code [here](https://gist.github.com/pk400/4366ed2cda0446c443bd4bdac7b71eb2). Basically we mount the decorator the way we want to:
```python
@tasks.loop(seconds=self.interval_seconds)
```
Becomes:
```python
self.scheduled_task = decorator(self.task)
```
This should properly setup our cog. The resulting bot would have this output:
```
initializing task A..
initializing task B..
initializing task C..
task is running with parameters task A 2
task is running with parameters task B 5
task is running with parameters task C 10
task is running with parameters task A 2
task is running with parameters task A 2
task is running with parameters task B 5
task is running with parameters task A 2
task is running with parameters task A 2
task is running with parameters task A 2
task is running with parameters task B 5
task is running with parameters task C 10
task is running with parameters task A 2
task is running with parameters task A 2
^C
```